### PR TITLE
Add new IHalDocumentMapper interface with additional context.

### LINKED ISSUE #6: Allow additional context to be provided to IHalDocumentMappers
When implementing IHalDocumentMapper, it is sometimes necessary to have access to additional information not available on the object being mapped.

The case I've encountered so far is when generating links, if something that needs to be in the Url is not available on the object. For example, if building a multi-tenanted API and the tenant Id is provided as part of the path, this will always be needed but won't normally be on the object.

The proposed solution is to provide an additional interface `IHalDocumentMapper<TResource, TContext` which allows the mapper to accept a context object as well as the resource to map.

### PR DESCRIPTION
**Pull request by jongeorge1**
Provides an additional `IHalDocumentMapper` interface that allows additional context to be supplied to a mapping operation. Background is described in Issue #6 .

Resolves #6.

### CODE REVIEW

**Comment by HowardvanRooijen** [comment]
Specs?

**Comment by idg10** [comment]
So TIL the the process of starting a review and clicking "yup, yup, ok, fine, yup" for each file makes it very easy to put the PR into an Approved state by accident. I have therefore just dismissed my own review. I didn't mention the need for specs because I knew @HowardvanRooijen had already mentioned that, so I ended up getting all the way through the review without opening any discussions. I was taking it as read that @HowardvanRooijen's point needed to be taken into account.

**Comment by jongeorge1** [comment]
I've added some specs around registration and initialisation. To start with I've added those that directly relate to the changes here, as well as a couple of other obvious bits that are as much about documentation as anything else - auditing and exception type registration.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            catch (TargetInvocationException ex)
+            {
+                // The exception will have been wrapped in a TargetInvocationException because we invoked via reflection, so we need to
+                // see if the inner exception was what we expected.
+                if (!(ex.InnerException is ArgumentException))
+                {
+                    throw;
+                }
+            }
+        }
```
Comment (idg10) [inline Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: This test (the one above this comment) has baffled me. Its name implies to me that, broadly speaking, we feed an exception into the system, and then we want to check that a particular status code will emerge from somewhere else in the system. I don't actually see where that check is happening—you don't seem to be asking the question "Is this the status code we expect?" Instead, on the face of it you seem to be passing the status code into the code under test.

My gut feeling here is that this is another scenario where we should really be testing that the outcome is what we want, but it seems like what we're actually doing is checking things slightly further upstream, and just presuming that if we set everything in motion in the right way, we'll get the outcome we want. This might be more fragile, and it's certainly harder to understand: I don't yet understand how this test tests what it says it tests.

What I _think_ I want (although it's possible I'd change my mind once I saw what was involved in practice) is to have something that actually throws the exception, invoke that in exactly the way Menes normally would and to check the outcome. After all, the point of these tests is to ask "Is the out of the box behaviour correct when we've configured this in the default way?" We're not shining a light on detailed little corner cases in the exception handling here, we're just verifying that the basic cases operate as intended, right? And in particular, the broader question being asked by this particular set of tests is "Does everything work right after we've configured it?" Those feel like exactly the cases where you do actually want to involve quite a lot of the pieces in a test, because one of the main things configuration influences is how all the moving parts get plugged together. The overall effect is the thing we're interested in at this point, I think.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+        [Then(@"an audit log builder service is added for auditing operations which return OpenApiResults")]
+        public void ThenAnAuditLogBuilderServiceIsAddedForAuditingOperationsWhichReturnOpenApiResults()
+        {
+            this.AssertServiceIsRegistered<IAuditLogBuilder, OpenApiResultAuditLogBuilder>();
```
Comment (idg10) [inline Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: Now this is much nicer. If your singleton tests earlier on in the file were are easy to read as these I'd be much happer.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        }
+
+        [Then(@"an audit log builder service is added for auditing operations which return a POCO")]
+        public void ThenAnAuditLogBuilderServiceIsAddedForAuditingOperationsWhichReturnAPOCO()
```
Comment (idg10) [inline Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: Not a major one this...but conventionally in .NET, acronyms get CamelCased, so POCO should become `Poco` in a method name.

File: Solutions/Menes.Specs/Steps/TestClasses/PetHalDocumentMapperWithContext.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// Copyright (c) Endjin. All rights reserved.
+// </copyright>
+
+namespace Menes.Specs.Steps.TestClasses
+{
+    using Menes.Hal;
+
+    public class PetHalDocumentMapperWithContext : IHalDocumentMapper<Pet, MappingContext>
+    {
+        public bool LinkMapConfigured { get; private set; }
```
Comment (idg10) [inline Solutions/Menes.Specs/Steps/TestClasses/PetHalDocumentMapperWithContext.cs]: It ever so slightly offends me that well over half of this class is identical to `PetHalDocumentMapper`. My inclination would have been to put the common bits in a common base class. And that's not so much to avoid duplication, as it is to make the meaningful differences stand out.

However I would concede that adding an extra moving part in the form of a base class increases complexity so reasonable people could disagree over whether it is a net win. My main reason for favouring this refactoring is that in practice I've often found that when looking at tests (particularly coming back to them when trying to understand existing functionality prior to augmenting or otherwise modifying it) I find it much easier to work out what's going on if test code focuses on what's distinctive about a particular scenario, instead of buring the one important detail in a cloud of incidental baggage.

File: Solutions/Menes.Hosting/Microsoft/Extensions/DependencyInjection/OpenApiHostingServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -166,6 +166,23 @@ public static class OpenApiHostingServiceCollectionExtensions
...
+        /// <typeparam name="TContext">The type of the additional context required by the HAL document mapper.</typeparam>
+        /// <typeparam name="TMapper">The type of the mapper.</typeparam>
+        /// <param name="services">The service collection to which to add the mapper.</param>
+        /// <returns>The service collection, configured with the HAL document mapper.</returns>
+        public static IServiceCollection AddHalDocumentMapper<TResource, TContext, TMapper>(this IServiceCollection services)
+            where TMapper : class, IHalDocumentMapper<TResource, TContext>
+        {
+            services.AddSingleton<TMapper>();
+            services.AddSingleton<IHalDocumentMapper>(s => s.GetRequiredService<TMapper>());
+            services.AddSingleton<IHalDocumentMapper<TResource, TContext>>(s => s.GetRequiredService<TMapper>());
```
[the remark this answers is not in the document]
Comment (jongeorge1) [reply Solutions/Menes.Hosting/Microsoft/Extensions/DependencyInjection/OpenApiHostingServiceCollectionExtensions.cs]: Doing it this way means that whichever of the three ways you resolve the mapper, you'll get the same instance.

File: Solutions/Menes.Specs/Features/OpenApiHostingInitialisation.feature [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	And an audit log builder service is added for auditing operations which return a POCO
+	And an audit log sink service is added for console logging
+	And auditing is enabled
+
+Scenario Outline: OpenApi host initialisation registers standard Menes exception types
+	Given I have created a service collection to register my services against
+	And I have added AspNetCore OpenApi hosting to the service collection
+	And I have built the service provider from the service collection
+	When I request an instance of the OpenApi host
+	Then the exception of type "<Exception Type>" is mapped to response code "<Mapped Response Code>"
```
[the remark this answers is not in the document]
Comment (jongeorge1) [reply Solutions/Menes.Specs/Features/OpenApiHostingInitialisation.feature]: That sounds like something we should make a hard decision on for our codebase... I'm happy to go with single quotes though.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            PetHalDocumentMapperWithContext mapper = provider.GetRequiredService<PetHalDocumentMapperWithContext>();
+            Assert.IsTrue(mapper.LinkMapConfigured);
+        }
+
+        [Then(@"the exception of type ""(.*)"" is mapped to response code ""(.*)""")]
+        public void ThenTheExceptionOfTypeIsMappedToResponseCode(string exceptionType, int statusCode)
+        {
+            IOpenApiExceptionMapper exceptionMapper = this.scenarioContext.Get<ServiceProvider>().GetRequiredService<IOpenApiExceptionMapper>();
+
+            // The only way to tell if it's been mapped without reflecting into the guts of the mapper is to try and register a new mapper for
```
[the remark this answers is not in the document]
Comment (jongeorge1) [reply Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: I didn't know that! Will do.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs (hunk 1 above)
Comment (jongeorge1) [reply Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: So, I think this is a combination of bad naming and (potentially) missing documentation.

I'm not at all convinced that we should be testing the overall way Menes handles exceptions. These tests are supposed to be about initialisation rather than what happens when operations are invoked and throw exceptions. Would you make the same argument for the auditing specs in this feature file? I.e. that in order to know auditing is configured in the way it should be, we should be calling into the Menes pipeline and ensuring that the result is logged?

I'm not saying I think we shouldn't have those tests somewhere - but I think there's still value in having these tests as they are (albeit with better naming). There'd ideally be a whole separate feature file for Menes exception handling (and one for auditing) that effectively documented how these things work. But the fact that specific exception handling behaviour is configured as part of this particular way of setting Menes up is separate from that.

And we definitely need to decide the terminology we want to use for this exception handling functionality. At the moment, to be consistent with the naming in the code, I'd rename this scenario to `OpenApi host initialisation maps standard Menes exception types to their corresponding HTTP status codes`.

File: Solutions/Menes.Specs/Steps/TestClasses/PetHalDocumentMapperWithContext.cs (hunk 4 above)
Comment (jongeorge1) [reply Solutions/Menes.Specs/Steps/TestClasses/PetHalDocumentMapperWithContext.cs]: I did consider a base class but it seemed like that would be equally confusing as now there'd be three things to keep in mind instead of two - to me, the naming was enough to express the difference. How strongly do you feel about this?

File: Solutions/Menes.Hosting/Microsoft/Extensions/DependencyInjection/OpenApiHostingServiceCollectionExtensions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (idg10) [reply Solutions/Menes.Hosting/Microsoft/Extensions/DependencyInjection/OpenApiHostingServiceCollectionExtensions.cs]: Ah, OK.

File: Solutions/Menes.Specs/Steps/TestClasses/PetHalDocumentMapperWithContext.cs (hunk 4 above)
Comment (idg10) [reply Solutions/Menes.Specs/Steps/TestClasses/PetHalDocumentMapperWithContext.cs]: Not very. If you're inclined not to change it having read my argument, then that's fair enough—there are pros and cons either way.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        {
+            this.scenarioContext.Get<ServiceCollection>().AddHalDocumentMapper<Pet, MappingContext, PetHalDocumentMapperWithContext>();
+        }
+
+        [Then(@"a service is added as a Singleton for type IOpenApiHost\{HttpRequest, IActionResult}")]
+        public void ThenAServiceIsAddedAsASingletonForTypeIOpenApiHostHttpRequestIActionResult()
+        {
+            ServiceDescriptor registration = this.scenarioContext.Get<ServiceCollection>().FirstOrDefault(x => x.ServiceType == typeof(IOpenApiHost<HttpRequest, IActionResult>));
+            Assert.NotNull(registration);
+            Assert.AreEqual(ServiceLifetime.Singleton, registration.Lifetime);
```
[the remark this answers is not in the document]
Comment (jongeorge1) [reply Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: Agreed.

File: Solutions/Menes.Specs/Features/OpenApiHostingInitialisation.feature [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+Scenario: Registering HAL document mappers with resource type parameters adds them to the container with the concrete type, the IHalDocumentMapper interface and the generic IHalDocumentMapper interface
+	Given I have created a service collection to register my services against
+	When I register a HalDocumentMapper for a resource type to the service collection
+	Then it should be added as a Singleton with the service type matching the concrete type of the mapper
+	And It should be added as a Singleton with a service type of IHalDocumentMapper
+	And it should be added as a Singleton with a service type of IHalDocumentMapper{TResource}
+
+Scenario: Registering HAL document mappers with resource and context type parameters adds them to the container with the concrete type, the IHalDocumentMapper interface and the generic IHalDocumentMapper interface
+	Given I have created a service collection to register my services against
```
[the remark this answers is not in the document]
Comment (jongeorge1) [reply Solutions/Menes.Specs/Features/OpenApiHostingInitialisation.feature]: I've done it via a Background block instead; I'm not sure that this particular thing will be needed outside this feature.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs (hunk 1 above)
Comment (idg10) [reply Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: > These tests are supposed to be about initialisation rather than what happens when operations are invoked and throw exceptions

Speaking as the named entity at the top of the feature file ("a developer" who wants "to be able to add Menes services and related components to my service collection"), I don't actually care a jot about how these configuration details effect the default behaviour. What I care about is the default behaviour. In fact I'd go as far as to say that one of the main points of having the various default configuration settings is to enable me to remain happily oblivious to how this aspects of configuration even work. I only want to have to think in terms of the relevant config settings if I want to change something.

So in my role as the alleged beneficiary of the feature file, I'd say that they are about how certain aspects of Menes behave out of the box. The key questions, then, are of the form: if I throw an `OpenApiBadRequestException` will I get a 400 response ? That's a condition and outcome that I care about. Config settings are basically irrelevant.

I've been thinking about this some more, and I've realised that *the heart of the matter* (for me) is this: suppose we changed the way the specified default behaviour was achieved, so that you get it through a total absence of config settings, and instead Menes just did the relevant thing by default unless you add config to change the behaviour. From the perspective of "default behaviour" that's an implementation detail. If I have a test that says "This is how this bit works by default" I wouldn't expect to need to change that test to handle this change from "defaults behaviour arises from automatically-supplied default config" to "default behaviour arises from how the system behaves when there's no config".

> Would you make the same argument for the auditing specs in this feature file? 

The only reason I didn't is that the tests stipulate that the auditing goes "to console" at which point this would then necessarily become a whole-process test. So I'm prepared to let that one go purely on pragmatic grounds, but I'll go on record as saying that it's a necessary evil rather than a lovely thing.

Having said all of that, we have made the decision that defaults are effected by supplying some initial configuration, and so it is arguably valid to have a test that just checks that this initial configuration looks like it's supposed to. I prefer the idea of a test that actually checks the behaviour because it seems inherently more robust, and a better expression of what I'm actually looking to get out of this as a developer using Menes, but if it's significantly more complex and the benefits are marginal, then deciding not to do that might be OK, in which case all we need to do is make it more clear what the test is actually checking for.

File: Solutions/Menes.Specs/Features/OpenApiHostingInitialisation.feature (hunk 9 above)
[the remark this answers is not in the document]
Comment (idg10) [reply Solutions/Menes.Specs/Features/OpenApiHostingInitialisation.feature]: Ah yes, I totally forgot about `Background:` - much simpler than my suggestion.

File: Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs (hunk 1 above)
Comment (jongeorge1) [reply Solutions/Menes.Specs/Steps/OpenApiHostingServiceCollectionExtensionsSteps.cs]: Possibly the disconnect here then is that I am viewing these more as unit tests, ensuring that a specific bit of the code (initialisation) behaves as expected. I think your concern would be met by adding integration tests that describe the expected out of the box behaviour, as well as how it can be modified during initialisation.

Which is effectively me saying "I'd like to leave these tests as they are" and we should add a work item to cover creating specs that describe and test the out of the box behaviour. Are you happy with that approach?
